### PR TITLE
Optimize cross cluster access role descriptor validation

### PR DESCRIPTION
**Pull request by n1v0lg**
This PR avoids an extra de-serialization of role descriptors received in a cross cluster access request, by pushing the validation down to the role building step (where we necessarily de-serialize the received role descriptors). 

This also has the effect that we return a `400` instead of a `401`. I could wrap the exception so that we return a `403` instead, but I think a `400` makes the most sense, since we received a bad payload. Currently, this failure is _not_ audited. I can add logic to detect it in [`authorize()`](https://github.com/elastic/elasticsearch/blob/b17dfc77b9c48313921aaafa9a9e3da3e2739fd8/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationService.java#L317) and emit an audit event, in a follow up, or in this PR. Just didn't want that to block review across time-zones.

### CODE REVIEW

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/AuthorizationServiceIntegTests.java [hunk 1]
```diff
@@ -121,6 +125,45 @@ public void testGetRoleDescriptorsIntersectionForRemoteCluster() throws IOExcept
         );
     }
 
+    public void testCrossClusterAccessWithInvalidRoleDescriptors() {
```
Comment (n1v0lg) [inline x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/AuthorizationServiceIntegTests.java:128]: This seemed like a high-level enough place to get (sort-of) end-to-end coverage on the failure case. Ideally, we'd instead do this against a cluster, with transport headers we set as part of the test, but I spent some time looking into that in the past and did not come up with a fully-working solution. Let me know if you prefer more in-depth coverage of this change.

**Comment by n1v0lg** [comment]
@elasticmachine run elasticsearch-ci/bwc

**Review by ywangd — COMMENTED** [review]
I have a comment on how we could optimize the new RoleReference and avoid de/serialization roundtrip.

> we return a 400 instead of a 401. I could wrap the exception so that we return a 403 instead, but I think a 400 makes the most sense, since we received a bad payload.

400 makes sense to me.

> this failure is not audited. I can add logic to detect it in [authorize()](https://github.com/elastic/elasticsearch/blob/b17dfc77b9c48313921aaafa9a9e3da3e2739fd8/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationService.java#L317) and emit an audit event, in a follow up, or in this PR.

Good point on audit. I haven't even thought about it. I prefer it to be a follow up.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/RoleReference.java [hunk 2]
```diff
@@ -136,15 +136,43 @@ public RoleKey id() {
...
-            return roleDescriptorsBytes;
+        public String getUserPrincipal() {
+            return userPrincipal;
         }
+    }
 
+    final class CrossClusterAccessInternalRoleReference extends BaseCrossClusterAccessRoleReference {
+
+        public CrossClusterAccessInternalRoleReference(CrossClusterAccessSubjectInfo.RoleDescriptorsBytes roleDescriptorsBytes) {
+            super(roleDescriptorsBytes);
```
[the remark this answers is not in the document]
Comment (n1v0lg) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/RoleReference.java:169]: I like your suggestion. I'll go ahead with implementing it. Overall, the change-set will be bigger but it's not too much effort, and it establishes a good re-usable pattern for the future.

**Comment by n1v0lg** [comment]
@elasticmachine update branch

**Comment by n1v0lg** [comment]
@elasticmachine run elasticsearch-ci/part-2-fips (fix coming in https://github.com/elastic/elasticsearch/pull/94978)

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/CrossClusterAccessAuthenticationService.java [hunk 3]
```diff
@@ -121,24 +113,6 @@ public AuthenticationService getAuthenticationService() {
...
-    private CrossClusterAccessSubjectInfo maybeRewriteForCrossClusterAccessUser(
-        AuthenticationService.AuditableRequest request,
-        CrossClusterAccessSubjectInfo crossClusterAccessSubjectInfo
-    ) {
-        final Authentication authentication = crossClusterAccessSubjectInfo.getAuthentication();
-        final Subject effectiveSubject = authentication.getEffectiveSubject();
-        final User user = effectiveSubject.getUser();
-        if (CrossClusterAccessUser.is(user)) {
-            logger.debug("Request [{}] performed by internal user [{}]. Will use pre-defined role descriptors", request, user.principal());
-            return CrossClusterAccessUser.subjectInfoWithRoleDescriptors(
```
Comment (n1v0lg) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/CrossClusterAccessAuthenticationService.java:133]: This had the side-effect that e.g., extra metadata was stripped. I think it makes sense to leave it out here, and instead sanitize all authentication instances the FC receives (esp. w.r.t. metadata). That way, we protect against bogus metadata more generally. WDYT?

Comment (n1v0lg) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/CrossClusterAccessAuthenticationService.java:133]: For sure, we can. It just keeps an extra layer of translation that makes the code slightly more complicated. I don't feel strongly though. Will add it back.

Comment (ywangd) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/CrossClusterAccessAuthenticationService.java:133]: Appologies, I think I misunderstood. You meant to drop it. I thought you wanted to keep it. My suggestion was to use `subjectInfoWithEmptyRoleDescriptors` if we are going to keep it. If you are going to drop it, my suggestion is invalid and I am OK with dropping it too. Sorry about the confusion.

Comment (n1v0lg) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/CrossClusterAccessAuthenticationService.java:133]: Yup sorry my intention was to drop it, and potentially add a more general sanitization step that removes "un-allow-listed" metadata from received authentication instances (in a separate PR, if at all).

**Comment by n1v0lg** [comment]
@elasticmachine run elasticsearch-ci/bwc

failure is unrelated and tracked here: https://github.com/elastic/elasticsearch/issues/95013

**Comment by n1v0lg** [comment]
@elasticmachine update branch

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/user/CrossClusterAccessUser.java [hunk 4]
```diff
@@ -56,23 +56,11 @@ public static boolean is(User user) {
...
-    public static CrossClusterAccessSubjectInfo subjectInfoWithEmptyRoleDescriptors(TransportVersion transportVersion, String nodeName) {
-        return subjectInfo(transportVersion, nodeName, RoleDescriptorsIntersection.EMPTY);
-    }
-
-    private static CrossClusterAccessSubjectInfo subjectInfo(
-        TransportVersion transportVersion,
-        String nodeName,
-        RoleDescriptorsIntersection roleDescriptorsIntersection
-    ) {
+    public static CrossClusterAccessSubjectInfo subjectInfo(TransportVersion transportVersion, String nodeName) {
```
Comment (ywangd) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/user/CrossClusterAccessUser.java:59]: Nit: can we have a brief comment here to explain why we always use empty role descriptor intersection?

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/authz/store/RoleReferenceTests.java [hunk 5]
```diff
@@ -68,13 +73,28 @@ public void testApiKeyRoleReference() {
...
+        final String source = "source";
+        final var fixedRoleReference = new RoleReference.FixedRoleReference(roleDescriptor, source);
+
+        final var future = new PlainActionFuture<RolesRetrievalResult>();
+        fixedRoleReference.resolve(mock(RoleReferenceResolver.class), future);
+        final var actualRetrievalResult = future.get();
+        assertThat(actualRetrievalResult.getRoleDescriptors(), equalTo(Set.of(roleDescriptor)));
+
+        final RoleKey roleKey = fixedRoleReference.id();
+        assertThat(roleKey.getNames(), hasItem(roleDescriptor.getName()));
```
Comment (ywangd) [inline x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/authz/store/RoleReferenceTests.java:94]: `hasItem` does not prevent `roleKey.getNames()` from having more items. So I suggest
```suggestion
        assertThat(roleKey.getNames(), containsInAnyOrder(roleDescriptor.getName()));
```

Alternatively, we can just compare `RoleKey` itself with
```
assertThat(roleKey, equalTo(new RoleKey(Set.of(roleDescriptor.getName()), source)));
```

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/AuthorizationServiceIntegTests.java [hunk 6]
```diff
@@ -121,6 +125,45 @@ public void testGetRoleDescriptorsIntersectionForRemoteCluster() throws IOExcept
...
+        final ThreadContext threadContext = internalCluster().getInstance(SecurityContext.class, nodeName).getThreadContext();
+        final AuthorizationService authzService = internalCluster().getInstance(AuthorizationService.class, nodeName);
+        final CrossClusterAccessSubjectInfo crossClusterAccessSubjectInfo = AuthenticationTestHelper.randomCrossClusterAccessSubjectInfo(
+            new RoleDescriptorsIntersection(
+                randomValueOtherThanMany(
+                    rd -> false == (rd.hasClusterPrivileges()
+                        || rd.hasApplicationPrivileges()
+                        || rd.hasConfigurableClusterPrivileges()
+                        || rd.hasRunAs()
+                        || rd.hasRemoteIndicesPrivileges()),
```
Comment (ywangd) [inline x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/AuthorizationServiceIntegTests.java:139]: Nit: can we reuse `RoleDescriptorStore.hasPrivilegesOtherThanIndex` here?

**Review by ywangd — APPROVED** [review]
LGTM 👍

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/AuthorizationServiceIntegTests.java (hunk 6 above)
Comment (n1v0lg) [reply x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/AuthorizationServiceIntegTests.java:139]: Yup. I want to make sure we're not relying on code to test itself though (currently nothing tests `hasPrivilegesOtherThanIndex`, other than the code in `AuthorizationServiceIntegTests`). To address this, I will:

1. Move `hasPrivilegesOtherThanIndex` directly to RoleDescriptor
2. Add a unit test for it
3. Re-use it here

**Pull request merged**